multi: Rework sync model to use hdr annoucements.

davecgh opened the pull request:
This overhauls the primary code that deals with synchronizing the blockchain with other peers on the network to use a model based on header announcements instead of inventory vectors.

Currently, all blocks are discovered via a combination of inventory vector announcements and the `getblocks` protocol message, both of which only include the hash of the blocks in question.

This is not ideal because it means blocks are blindly downloaded based on those announcements without knowing if they're actually likely to be something that is useful since there is no further information such as what blocks they connect to.  It also means that extra logic is needed to deal with orphan blocks (those whose parents are not yet known) such as caching them, determining known orphan roots, and expiration.

In short, the current method generally ends up wasting bandwidth and also makes it much more difficult to detect certain classes of malicious behavior.

The recently added capability of blockchain to process headers independently from blocks while the block data is added out of order later opened the door for a much better model that addresses all of the aforementioned issues as well as paves the way for many other related enhancements.

The new model discovers and downloads all of the headers prior to any block data via the `getheaders` protocol message and then uses those headers to determine exactly which blocks need to be downloaded to reach the tip of the best chain.  Notably, this means orphan blocks are now a thing of the past as blocks that do not connect are no longer downloaded under any circumstance.

Further, the new model also makes use of `sendheaders` so that all block announcements are made via the associated block header.  This in turn is used to better determine if an announced block is likely to be useful prior to downloading it.

It should be noted that the changes herein are intentionally limited to those necessary to use the new sync model based on headers in an incremental fashion to help simplify review and make it easier to assert correctness.  There are many more improvements that this model paves the way to support planned for future commits.  For example, syncing from multiple peers in parallel and improved DoS protection.

All of the changes are split up into a series of logic commits that describe each step to help ease the review process.

### High-level overview of the key features:

- All headers are downloaded and examined prior to downloading any blocks
- The initial header sync process:
  - Detects and recovers from stalled/malicious peers
  - Includes periodic logging with cumulative stats and a progress percentage
- The concept of orphan blocks no longer exists
  - This means blocks which are not already known to connect are not downloaded
- All block announcements are handled via header announcements
  - Detects and prevents malicious behavior related to orphan headers
- The chain sync process:
  - Starts once the headers are downloaded and entails both downloading blocks as well as verifying them
  - Uses the headers to determine the best blocks to download
  - Pipelines the requests to increase throughput
  - Actively avoids downloading duplicate blocks
  - Logs the overall verification progress versus the timestamp of the last processed block
- The sync height is dynamically updated as new headers are discovered
- The logging behavior is changed once steady state mode is reached to instead log new blocks along with their hashes and additional stats immediately

---

This is work towards #1145.

Fixes #2252.

davecgh commented on internal/netsync/manager.go:
```diff
@@ -919,98 +873,181 @@ func (m *SyncManager) handleHeadersMsg(hmsg *headersMsg) {
+	// headers and request any headers starting from the best header the local
+	// chain knows in order to (hopefully) discover the missing headers.
+	//
+	// Meanwhile, also keep track of how many many times the peer has
```
```suggestion
	// Meanwhile, also keep track of how many times the peer has
```

davecgh replied:
Removed.

davecgh commented on internal/progresslog/logger.go:
```diff
@@ -81,14 +81,14 @@ func (l *Logger) LogProgress(block *wire.MsgBlock, forceLog bool) {
 	// Log information about chain progress.
 	l.subsystemLogger.Infof("%s %d %s in the last %s (%d %s, %d %s, %d %s, "+
-		"%d %s, height %d, %s)", l.progressAction,
+		"%d %s, height %d, progress %0.2f%%)", l.progressAction,
```
This probably should only conditionally log this when the `forceLog` parameter is not set. Otherwise, it shows `progress 100%` with every new block that shows up after catching up.

davecgh replied:
I decided to handle this directly in `netsync` instead.  The code will now determine when the chain has reached steady state (is current and was already current before) and modify the behavior to log new blocks as they are connected along with their hash and additional details.  This provides a cleaner distinction between the initial verification and catchup phase and steady state operation.

davecgh commented:
I added an additional commit since the original PR was opened to change the verification progress reported by the RPC server for the new model as well.

matheusd commented on blockchain/chainquery.go:
```diff
@@ -154,3 +154,80 @@ func (b *BlockChain) BestInvalidHeader() chainhash.Hash {
+		return nil
+	}
+
+	// No blocks are needed from branches that do not lead to the current best
+	// header.
+	tip := b.bestChain.Tip()
+	b.index.RLock()
+	bestHeader := b.index.bestHeader
+	if bestHeader.Ancestor(tip.height) != tip {
+		b.index.RUnlock()
+		return nil
+	}
```
This doesn't seem to be the case, since while doing a simnet test, reorgs are working just fine so I'm misunderstanding something, just not clear yet what exactly.

davecgh replied:
You're right.  This is actually some code from the multi-download that is under development too that snuck in when I was splitting this bit off.  Nice catch!  I'll get this updated here in a bit.

The reason reorgs are still working for you as expected is because there are currently two paths.  One is the catchup path, which is this one.  The other one is the immediate download path that happens once you are current for better throughput, mainly for votes at the current time.

davecgh replied:
Updated.

matheusd reviewed (commented):
(still doing some functional testing)

Got the following output while syncing on mainnet:

```
2021-01-18 16:03:03.837 [INF] SYNC: New valid peer [x] (outbound) (/dcrwire:0.4.0/dcrd:1.6.0(pre)/)
2021-01-18 16:03:12.919 [INF] SYNC: Processed 258000 headers in the last 10.15s (progress 94.02%)
2021-01-18 16:03:13.521 [INF] SYNC: Syncing headers to block height 521071 from peer [y] (outbound)
2021-01-18 16:03:19.309 [INF] SYNC: Processed 133074 headers in the last 6.38s (progress 100.00%)
2021-01-18 16:03:19.309 [INF] SYNC: Headers synced -- downloading blocks
2021-01-18 16:03:19.711 [INF] SYNC: Processed 2000 headers in the last 400ms (progress 100.00%)
2021-01-18 16:03:20.191 [INF] SYNC: Processed 2000 headers in the last 470ms (progress 100.00%)
2021-01-18 16:03:20.293 [INF] SYNC: Processed 2000 headers in the last 100ms (progress 100.00%)
2021-01-18 16:03:20.840 [INF] SYNC: Processed 2000 headers in the last 540ms (progress 100.00%)
2021-01-18 16:03:20.866 [INF] SYNC: Processed 2000 headers in the last 20ms (progress 100.00%)
2021-01-18 16:03:21.314 [INF] SYNC: Processed 2000 headers in the last 440ms (progress 100.00%)
2021-01-18 16:03:21.441 [INF] SYNC: Processed 2000 headers in the last 120ms (progress 100.00%)
2021-01-18 16:03:21.787 [INF] SYNC: Processed 2000 headers in the last 340ms (progress 100.00%)
2021-01-18 16:03:22.015 [INF] SYNC: Processed 2000 headers in the last 220ms (progress 100.00%)
2021-01-18 16:03:22.268 [INF] SYNC: Processed 2000 headers in the last 250ms (progress 100.00%)
2021-01-18 16:03:22.591 [INF] SYNC: Processed 2000 headers in the last 320ms (progress 100.00%)
2021-01-18 16:03:22.747 [INF] SYNC: Processed 2000 headers in the last 150ms (progress 100.00%)
2021-01-18 16:03:23.164 [INF] SYNC: Processed 2000 headers in the last 410ms (progress 100.00%)
```

davecgh commented on internal/netsync/manager.go:
```diff
@@ -819,15 +824,37 @@ func (m *SyncManager) handleBlockMsg(bmsg *blockMsg) {
 	}
 
-	// Log information about the block and update the chain state.
+	// Log information about the block.  Use the progres logger when the chain
```
Modified.

davecgh commented on internal/netsync/manager.go:
```diff
@@ -919,97 +873,190 @@ func (m *SyncManager) handleHeadersMsg(hmsg *headersMsg) {
+			m.hdrSyncState.headersSynced = headersSynced
+			m.hdrSyncState.stopStallTimeout()
+
+			log.Info("Headers synced -- downloading blocks")
```
Added.  Also added the best chain tip hash and height to the one that is printed when the chain is considered synced.

davecgh commented:
> 
> 
> Got the following output while syncing on mainnet:
> 
> ```
>...
> 2021-01-18 16:03:21.314 [INF] SYNC: Processed 2000 headers in the last 440ms (progress 100.00%)
> ```

That is fixed with the latest.

matheusd commented:
Woot :P

While downloading headers: 
```
$ dcrctl getblockchaininfo
{
  "chain": "mainnet",
  "blocks": 0,
  "headers": 182000,
  "syncheight": 521083,
  "bestblockhash": "298e5cc3d985bfe7f81dc135f360abe089edd4396b86d2de66b0cef42b21d980",
  "difficulty": 453115903,
  "difficultyratio": 32767.74999809,
  "verificationprogress": 0,
  "chainwork": "0000000000000000000000000000000000000000000000000000800040002000",
  "initialblockdownload": true,
  "maxblocksize": 393216,
  "deployments": {
  ...
```

And then while downloading blocks:

```  
$ dcrctl getblockchaininfo
  {
  "chain": "mainnet",
  "blocks": 1472,
  "headers": 521084,
  "syncheight": 521084,
  "bestblockhash": "000000000000157f9f24f293d8153b8de0465e8558f2c386f080047b67d1502c",
  "difficulty": 442693272,
  "difficultyratio": 169527.16125555,
  "verificationprogress": 0.002824880441541095,
  "chainwork": "0000000000000000000000000000000000000000000000000b668d107cb857b2",
  "initialblockdownload": true,
  "maxblocksize": 393216,
  "deployments": {
  ...
```

matheusd reviewed (approved):
All issues resolved, tested reorgs on the standard simnet setup and syncing in mainnet.

Progressing nicely :P

davecgh commented:
I updated the initial chain synced log message to also account for the case when the chain is already fully synced after the initial header sync process so that it prints immediately as expected versus waiting until a new block shows up.

 [Changes](https://github.com/decred/dcrd/compare/194420bc792487fe9e91a4f970c14decf2868bad..2e55226842c5e40e481ddadc57ad011511a4c22b).

rstaudt2 reviewed (commented):
Still reviewing/testing but I did notice one strange behavior.  It's intermittent but about 50% of the time the header progress message just keeps getting logged non-stop after the initial header sync is completed (with a fresh db on mainnet):

EDIT: Looks like this was already mentioned above as resolved, but it appears to still be occurring intermittently.
```
2021-01-18 20:57:49.085 [INF] CMGR: 16 addresses found from seeder mainnet-seed.dcrdata.org
2021-01-18 20:57:49.107 [INF] CMGR: 16 addresses found from seeder mainnet-seed-1.decred.org
2021-01-18 20:57:49.771 [INF] CMGR: 16 addresses found from seeder mainnet-seed-2.decred.org
2021-01-18 20:57:50.065 [INF] CMGR: 16 addresses found from seeder mainnet-seed.planetdecred.org
2021-01-18 20:57:50.478 [INF] CMGR: 16 addresses found from seeder mainnet-seed.jamieholdstock.com
2021-01-18 20:57:53.902 [INF] SYNC: New valid peer 107.152.35.163:9108 (outbound) (/dcrwire:0.4.0/dcrd:1.5.1/)
2021-01-18 20:57:53.902 [INF] SYNC: Syncing headers to block height 521176 from peer 107.152.35.163:9108 (outbound)
2021-01-18 20:57:54.128 [INF] SYNC: New valid peer 51.195.222.207:9108 (outbound) (/dcrwire:0.4.0/dcrd:1.6.0/)
2021-01-18 20:57:54.128 [INF] SYNC: New valid peer 18.198.44.215:9108 (outbound) (/dcrwire:0.4.0/dcrd:1.5.2/)
2021-01-18 20:57:54.152 [INF] SYNC: New valid peer 206.81.16.204:9108 (outbound) (/dcrwire:0.4.0/dcrd:1.7.0(pre)/)
2021-01-18 20:57:54.335 [INF] SYNC: New valid peer 104.199.150.126:9108 (outbound) (/dcrwire:0.4.0/dcrd:1.5.2/)
2021-01-18 20:57:54.335 [INF] SYNC: New valid peer 139.99.135.146:9108 (outbound) (/dcrwire:0.4.0/dcrd:1.6.0/)
2021-01-18 20:57:55.150 [INF] SYNC: New valid peer [2a0b:b7c4:1:10c1::1]:9108 (outbound) (/dcrwire:0.4.0/dcrd:1.6.0(rc4)/)
2021-01-18 20:57:55.317 [INF] SYNC: New valid peer [2001:19f0:7001:3511:5400:1ff:feb3:3d18]:9108 (outbound) (/dcrwire:0.4.0/dcrd:1.5.1/)
2021-01-18 20:57:58.976 [INF] SYNC: Processed 52000 headers in the last 10.12s (progress 9.99%)
2021-01-18 20:58:09.059 [INF] SYNC: Processed 114000 headers in the last 10.08s (progress 30.34%)
2021-01-18 20:58:19.672 [INF] SYNC: Processed 428000 headers in the last 10.61s (progress 45.68%)
2021-01-18 20:58:29.748 [INF] SYNC: Processed 540000 headers in the last 10.07s (progress 70.57%)
2021-01-18 20:58:39.785 [INF] SYNC: Processed 302000 headers in the last 10.03s (progress 87.86%)
2021-01-18 20:58:46.313 [INF] SYNC: Processed 149177 headers in the last 6.52s (progress 100.00%)
2021-01-18 20:58:46.313 [INF] SYNC: Initial headers sync complete -- downloading blocks (best header hash 000000000000000006767d42fc60cf1dba3e3ca83075a51514ff25de7d2dbe3b, height 521177)
2021-01-18 20:58:46.323 [INF] SYNC: Processed 2000 headers in the last 10ms (progress 100.00%)
2021-01-18 20:58:46.347 [INF] SYNC: Processed 2000 headers in the last 20ms (progress 100.00%)
2021-01-18 20:58:46.468 [INF] SYNC: Processed 2000 headers in the last 120ms (progress 100.00%)
2021-01-18 20:58:46.518 [INF] SYNC: Processed 2000 headers in the last 50ms (progress 100.00%)
2021-01-18 20:58:46.604 [INF] SYNC: Processed 2000 headers in the last 80ms (progress 100.00%)
2021-01-18 20:58:46.713 [INF] SYNC: Processed 2000 headers in the last 100ms (progress 100.00%)
2021-01-18 20:58:46.767 [INF] SYNC: Processed 2000 headers in the last 50ms (progress 100.00%)
2021-01-18 20:58:46.921 [INF] SYNC: Processed 2000 headers in the last 150ms (progress 100.00%)
2021-01-18 20:58:46.967 [INF] SYNC: Processed 2000 headers in the last 40ms (progress 100.00%)
2021-01-18 20:58:47.175 [INF] SYNC: Processed 2000 headers in the last 200ms (progress 100.00%)
2021-01-18 20:58:47.187 [INF] SYNC: Processed 2000 headers in the last 10ms (progress 100.00%)
2021-01-18 20:58:47.228 [INF] SYNC: Processed 2000 headers in the last 40ms (progress 100.00%)
2021-01-18 20:58:47.238 [INF] SYNC: Processed 2000 headers in the last 10ms (progress 100.00%)
2021-01-18 20:58:47.427 [INF] SYNC: Processed 2000 headers in the last 180ms (progress 100.00%)
...
```

davecgh commented:
> 
> 
> Still reviewing/testing but I did notice one strange behavior. It's intermittent but about 50% of the time the header progress message just keeps getting logged non-stop after the initial header sync is completed (with a fresh db on mainnet):
> 
> EDIT: Looks like this was already mentioned above as resolved, but it appears to still be occurring intermittently.
> 

Upon closer inspection I see this is something else than what I thought it was.  The culprit was a new block showing up during the initial headers sync.  I've pushed a proper fix for it.

davecgh commented on internal/progresslog/logger.go:
```diff
@@ -0,0 +1,135 @@
+	}
+}
+
+// LogProgress details for the provided block and periodically (every 10
```
Added.

davecgh commented on internal/progresslog/logger.go:
```diff
@@ -0,0 +1,135 @@
+	l.subsystemLogger.Infof("%s %d %s in the last %s (%d %s, %d %s, %d %s, "+
+		"%d %s, height %d, progress %0.2f%%)", l.progressAction,
+		l.receivedBlocks, pickNoun(l.receivedBlocks, "block", "blocks"),
+		duration.Truncate(time.Millisecond*10),
```
I personally kind of prefer the way the built-in duration stringer changes to e.g. `220 ms` when necessary versus `0.22s`.  However, since the predominant time scale is seconds, I went ahead and changed this (as well as the case for blocks) per your suggestion.

davecgh commented on internal/netsync/manager.go:
```diff
@@ -178,14 +184,71 @@ type syncMgrPeer struct {
-	expiration time.Time
+// stopStallTimeout stops the progress stall timer while ensuring to read from
+// the timer's channel in the case the timer already expired which can happen
+// due to the fact the stop happens in between channel reads.   This behavior is
```
Updated.

davecgh commented on internal/netsync/manager.go:
```diff
@@ -178,14 +184,71 @@ type syncMgrPeer struct {
+
+// resetStallTimeout resets the progress stall timer while ensuring to read from
+// the timer's channel in the case the timer already expired which can happen
+// due to the fact the reset happens in between channel reads.   This behavior
```
Updated.

davecgh commented on internal/netsync/manager.go:
```diff
@@ -310,16 +321,60 @@ func chainBlockLocatorToHashes(locator blockchain.BlockLocator) []chainhash.Hash
 	}
 
+	chain := m.cfg.Chain
 	best := m.cfg.Chain.BestSnapshot()
```
Good call.  Things have moved around a bit as and I missed updating that obviously.  Updated this along with another case.

davecgh commented on internal/netsync/manager.go:
```diff
@@ -363,62 +418,53 @@ func (m *SyncManager) startSync() {
+		}
 	} else {
 		log.Warnf("No sync peer candidates available")
 	}
```
Indeed.  Done.

rstaudt2 commented:
> Upon closer inspection I see this is something else than what I thought it was. The culprit was a new block showing up during the initial headers sync. I've pushed [a proper fix for it](https://github.com/decred/dcrd/compare/2e55226842c5e40e481ddadc57ad011511a4c22b..23fa9f0380446a27881b8931fac3d6b7673d9f6c).

I tested a bunch more and can confirm that this is resolved.

rstaudt2 reviewed (commented):
This looks great.  I did a bunch more testing, and everything has been working as expected.  I just had a few typos called out inline.

dnldd reviewed (approved):
Looks good.

davecgh commented on internal/netsync/manager.go:
```diff
@@ -49,6 +48,21 @@ const (
+	maxConsecutiveOrphanHeaders = 10
+
+	// headerSyncStallTimeoutSecs is the number of seconds to wait for progress
+	// during the header sync progress before stalling the sync and
```
Updated.

davecgh commented on internal/netsync/manager.go:
```diff
@@ -1153,97 +971,226 @@ func (m *SyncManager) handleHeadersMsg(hmsg *headersMsg) {
+				newBestHeaderHeight)
+			m.progressLogger.SetLastLogTime(time.Now())
+
+			// Potentially update whether the chain believe it is current now
```
Modified.

davecgh commented on internal/netsync/manager.go:
```diff
@@ -1300,14 +1241,14 @@ func (m *SyncManager) needTx(hash *chainhash.Hash) bool {
 
-// handleInvMsg handles inv messages from all peers.
-// We examine the inventory advertised by the remote peer and act accordingly.
+// handleInvMsg handles inv messages from all peers.  This entails examing the
```
Corrected.

davecgh commented on internal/progresslog/logger.go:
```diff
@@ -0,0 +1,135 @@
+	}
+
+	// Log information about chain progress.
+	l.subsystemLogger.Infof("%s %d %s in the last %0.2fs (%d %s, %d %s, %d %s,"+
```
Added.

davecgh commented:
In addition to addressing all of the review items, I added the interval since the previous block to the "New block ...." log print stats once steady state is reached in the latest update.
